src/stockpile.clj

```diff
@@ -0,0 +1,364 @@
+
+(def portable-path-bytes (set (map int portable-path-chars)))
+
+(defn encode-portable-path-element [s]
+  "Encodes the string s for use as a component in a portable filesystem path.
+  The current encoding is very similar to URL percent encoding, but
+  with an even more limited set of pass-through characters."
+  (let [utf-8 (.getBytes (str/replace s "%" "%%") "UTF-8")]
+    (String. (byte-array
+              (mapcat (fn [b]
+                        (if (portable-path-bytes b)
+                          [b]
+                          (map int (format "%%%02x" b))))
+                      utf-8))
+             "UTF-8")))
+
+(defn decode-portable-path-element [encoded]
+  (let [n (count encoded)
+        utf-8 (byte-array
+               (loop [i 0
+                      result '()]
+                 (if-let [c (nth encoded i nil)]
+                   (cond
+                     (= c \%)
+                     (if (= \% (nth encoded (inc i) nil))
+                       (recur (+ i 2) (cons (int \%) result))
+                       (if (>= (+ i 2) n)
+                         (throw (IllegalArgumentException.
+                                 (str "unexpected end of encoded string "
+                                      (pr-str encoded))))
+                         ;; May throw NumberFormatException
+                         (recur (+ i 3)
+                                (cons (Integer/parseInt (subs encoded
+                                                              (+ i 1)
+                                                              (+ i 3))
+                                                        16)
+                                      result))))
+
+                     (portable-path-bytes (int c))
+                     (recur (inc i) (conj result (int c)))
+
+                     :else
+                     (-> "invalid character \"%c\" at position %d in %s"
+                         (format c i encoded)
+                         IllegalArgumentException.
+                         throw))
+                   (reverse result))))]
+    (String. utf-8 "UTF-8")))
+
+(defn- parse-integer [x]
+  (try
+    (Long/parseLong x)
+    (catch NumberFormatException ex
+      nil)))
+
+(defprotocol AsPath
+  (as-path [x]))
+
+(extend-protocol AsPath
+  Path
+  (as-path [x] x)
+  String
+  (as-path [x] (-> (path-get x)))
+  File
+  (as-path [x] (.toPath x)))
+
+(defprotocol Entry
+  (entry-id [entry])
+  (entry-meta [entry]))
+
+(defrecord MetaEntry [id meta]
+  Entry
+  (entry-id [this] id)
+  (entry-meta [this] meta))
+
+(extend-protocol Entry
+  Long
+  (entry-id [this] this)
+  (entry-meta [this] nil)
+  BigInt
+  (entry-id [this] (long this))
+  (entry-meta [this] nil)
+  BigInteger
+  (entry-id [this] (long this))
+  (entry-meta [this] nil))
+
+(defn entry [id metadata]
+  (let [id (if-not (integer? id)
+             (throw
+              (IllegalArgumentException. (str "id is not an integer: " id)))
+             (cond
```

Comment:
This was also trying to make sure the id was reduced as much as possible since (in some cases) it looks like clojure doesn't do that itself.  i.e. wanted 5N -> 5.